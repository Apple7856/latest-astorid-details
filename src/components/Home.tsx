import { Box, Button, Container, TextField } from '@mui/material'
import axios from 'axios'
import React, { useEffect, useState } from 'react'
import { useNavigate } from 'react-router-dom'

type StateData = {
    name: string
    nasa_jpl_url: string
    is_potentially_hazardous_asteroid: boolean
}

const Home = () => {

    const [data, setData] = useState<StateData>({
        name: "",
    nasa_jpl_url: "",
    is_potentially_hazardous_asteroid: true,
    });
    const [done, setDone] = useState(false);
    const [inputData, setInputData] = useState<number>(0);
    const [disabledButton, setDisabledButton] = useState<boolean>(false);

    const navigate = useNavigate();

    function fetchSearchData() {
        axios({
            method: "GET",
            url: `https://api.nasa.gov/neo/rest/v1/neo/${inputData}?api_key=wUyFxo6gsWNWArjWdwelzSDXcqRkYCajzdnYSkoF`,
            responseType: 'json'
        })
            .then((result) => {
                setData({
                    name: result.data.name,
                    nasa_jpl_url: result.data.nasa_jpl_url,
                    is_potentially_hazardous_asteroid: result.data.is_potentially_hazardous_asteroid
                })
                setDone(true);
            })
            .catch((error) => {
                console.log(error);
            })
    }

    const fetchRandomData = () => {
        setDisabledButton(true);
        axios({
            method: "GET",
            url: "https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=wUyFxo6gsWNWArjWdwelzSDXcqRkYCajzdnYSkoF",
            responseType: 'json'
        })
            .then((responce) => {
                const astoridId = responce.data.near_earth_objects[Math.floor(Math.random() * responce.data.near_earth_objects.length) + 1].id;
                axios({
                    method: "GET",
                    url: `https://api.nasa.gov/neo/rest/v1/neo/${astoridId}?api_key=wUyFxo6gsWNWArjWdwelzSDXcqRkYCajzdnYSkoF`,
                    responseType: 'json'
                })
                    .then((result) => {
                        setData({
                            name: result.data.name,
                            nasa_jpl_url: result.data.nasa_jpl_url,
                            is_potentially_hazardous_asteroid: result.data.is_potentially_hazardous_asteroid
                        })
                        setDone(true);
                    })
                    .catch((error) => {
                        console.log(error);
                    })
            })
            .catch((error) => {
                console.log(error);
            })
    }
    useEffect(() => {
        if (done) {
            const id = navigate("/details", { state: data });
            return ()=>{
                setDone(false);
            }
        }
    }, [data])
    
    return (
        <Container maxWidth="md" sx={{ backgroundColor: "#b3ffb3", width: "100%", height: "100vh", display: "flex", flexDirection: "column", alignItems: "center", justifyContent: "center" }}>
            <Box sx={{ display: "flex", alignItems: "center", width: "350px", justifyContent: "space-between" }}>
                <TextField data-testid="inputfield" color='secondary' type="number" onChange={(e:any)=>setInputData(e.target.value)} />
                <Button variant='contained' data-testid="searchButton" disabled={!inputData || disabledButton} onClick={()=>fetchSearchData()}>Search</Button>
            </Box>
            <Button variant='contained' data-testid="randomButton" color='secondary' disabled={disabledButton} sx={{ marginTop: "30px" }} onClick={() => fetchRandomData()}>Random Search</Button>
        </Container>
    )
}

export default Home;