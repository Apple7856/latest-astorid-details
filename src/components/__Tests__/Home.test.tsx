import { act, cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import Home from '../Home';
import axios from 'axios';

beforeEach(() => {
    document.body.innerHTML = "";
});

afterEach(() => {
    cleanup();
});

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom') as any,
    useNavigate: () => mockedUsedNavigate,
}));


describe("Random Button Test", () => {
    test("Random Button Enable", () => {
        const component = render(<Home />);
        const randomButton = component.getByRole('button', { name: /Random Search/i });
        expect(randomButton).toBeEnabled();
    })
    test("Random Button Disabled", async () => {
        await act(async () => {
            const component = render(<Home />);
            const randomButton = component.getByRole('button', { name: /Random Search/i });
            // const inputfield: any = component.getByTestId('inputfield').querySelector('input');
            await fireEvent.click(randomButton);
            expect(randomButton).toBeDisabled();
        })
    })
})

describe("Home Component Test", () => {

    test("Input Field Test", async () => {
        await act(async () => {
            const component = render(<Home />);
            const inputfield = component.getByTestId('inputfield');
            expect(inputfield).toBeTruthy();
        })
    })
    test("Search Button Test", () => {
        const component = render(<Home />);
        const searchButton = component.getByTestId('searchButton');
        expect(searchButton).toBeTruthy();
    })
    test("Random Button Test", () => {
        const component = render(<Home />);
        const randomButton = component.getByTestId('randomButton');
        expect(randomButton).toBeTruthy();
    })
    // test("Create SnapShot", () => {
    //     const tree = renderer.create(<Home />);
    //     expect(tree).toMatchSnapshot();
    // })

})

describe("Api Test When Astroid Id is not given", () => {
    
    // const mockData = { data: {} };
    // const getAstoridId = jest.fn((url) => {
    //     return mockData
    // });
    // test("Return all Astroid details", () => {
    //     expect(getAstoridId(mockUrl)).toBe(mockData);
    // })
    // test("Return Astroid url", () => {
    //     expect(getAstoridId).toBeCalledWith(mockUrl);
    // })
    test('API REsponse status test', async () => {
        jest.useFakeTimers()
        const response = await axios.get('https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=wUyFxo6gsWNWArjWdwelzSDXcqRkYCajzdnYSkoF');
        setTimeout(() => {
            expect(response.status).toBe(200)
        }, 60000)
        jest.runAllTimers()
    })
})

describe("Api Test When Astroid Id is given", () => {
    const mockUrl = 'https://api.nasa.gov/neo/rest/v1/neo/2001943?api_key=wUyFxo6gsWNWArjWdwelzSDXcqRkYCajzdnYSkoF';
    const mockData = { data: {} };
    const getAstoridDetails = jest.fn((url) => {
        return mockData
    });
    test("Return Astroid Id details", () => {
        expect(getAstoridDetails(mockUrl)).toBe(mockData);
    })
    test("Return Astroid Id details url", () => {
        expect(getAstoridDetails).toBeCalledWith(mockUrl);
    })
})