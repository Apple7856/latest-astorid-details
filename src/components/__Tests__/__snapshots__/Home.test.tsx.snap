// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Test Create SnapShot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthMd css-1n2gf07-MuiContainer-root"
>
  <div
    className="MuiBox-root css-10xrxdt"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="inputfield"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorSecondary MuiInputBase-formControl css-hjtp1-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="number"
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-nnbavb"
          >
            <span
              className="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="searchButton"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Search
    </button>
  </div>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zfu506-MuiButtonBase-root-MuiButton-root"
    data-testid="randomButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Random Search
  </button>
</div>
`;
